### PR TITLE
Handle some invalid parameters for save files and calculate file size

### PR DESCRIPTION
**Pull request by jjsat**
This PR adds some checks for invalid parameters to the savegame functions and adds calculation of the save game total size.

Instead of taking out the savegame code some demos seem to simply provide invalid parameters so that the save function errors out. This PR fixes a "savegame corrupted" message in the Alien Breed: Impact demo and a null-pointer crash on startup for the Alone in the Dark demo, which passes a null filename.

Deux Ex HR DC enumerates the save games on startup and checks for a certain minimum file size. With this PR loading a savegame works because the savegames are no longer flagged as invalid by the game (change to savedata_get_list_item()).

The lines like "****** sysutil savedata parameter error : xx ******" comments are what the PS3 savedata sysutil prints to tty on the specific error.

I was sitting on these changes for a while so might as well create a PR :-)

### CODE REVIEW

**Comment by MarioSonic2987** [comment]
image
Fixes GTA IV.

**Comment by LuismaSP89** [comment]
Sorry, but could this PR solve/can handle this (https://github.com/RPCS3/rpcs3/issues/3767) ? or is totally a different solution for other games like you said (Alien Breed, Alone in the dark, etc)

Thanks in advance.

**Comment by jjsat** [comment]
@LuismaSP89 You can try the appveyor artifact and see if it fixes your issue (https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.4-7182/artifacts)

**Comment by SakataGintokiYT** [comment]
bez tytulu




```
E {PPU[0x1000000] Thread (main_thread) [0x01ae40f8]} cellSaveData: cellSaveDataEnableOverlay(enable=0)
E {PPU[0x100001a] Thread (SaveData) [0x01ae41a8]} cellSaveData: cellSaveDataUserAutoSave(version=0, userId=1, dirName=“BLES01069-PLAY”, errDialog=0, setBuf=*0xd017f9fc, funcStat=*0x1be6b78, funcFile=*0x1be6b80, container=0xffffffff, userdata=*0x23aedb40)
E {PPU[0x100001a] Thread (SaveData) [0x01f1fee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, “/dev_hdd0/game/BLES01069/USRDIR/UNREALENGINE3/BINARIES/PS3/SAVE/ICON1.PAM” [1]
E {PPU[0x100001a] Thread (SaveData) [0x01f1fee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, “/dev_hdd0/game/BLES01069/USRDIR/UNREALENGINE3/BINARIES/PS3/SAVE/PIC1.PNG” [1]
E {PPU[0x100001a] Thread (SaveData) [0x01f1fee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, “/dev_hdd0/game/BLES01069/USRDIR/UNREALENGINE3/BINARIES/PS3/SAVE/SND0.AT3” [1]
```

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1543864/RPCS3.log.gz)

**Comment by LuismaSP89** [comment]
@jjsat Great! This PR solves the problem with bomberman savegames ; ) thank you so much.

**Comment by jjsat** [comment]
@SakataGintokiYT I guess the game freezes at this point? Is this a new error with this PR or does this also happen on master?

**Comment by greentop** [comment]
This PR corrects the "Save New Entry" issue in bug #3747.

**Comment by SakataGintokiYT** [comment]
@jjsat 
No, this game doesn't hang, but it can't go further and when I press the X button then emulator spam these lines.
Yes, I have the same problem in the master build.

```
E {PPU[0x1000000] Thread (main_thread) [0x01ae40f8]} cellSaveData: cellSaveDataEnableOverlay(enable=0)
E {PPU[0x100001a] Thread (SaveData) [0x01ae41a8]} cellSaveData: cellSaveDataUserAutoSave(version=0, userId=1, dirName=“BLES01069-PLAY”, errDialog=0, setBuf=*0xd017f9fc, funcStat=*0x1be6b78, funcFile=*0x1be6b80, container=0xffffffff, userdata=*0x23aedb40)
E {PPU[0x100001a] Thread (SaveData) [0x01f1fee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, “/dev_hdd0/game/BLES01069/USRDIR/UNREALENGINE3/BINARIES/PS3/SAVE/ICON1.PAM” [1]
E {PPU[0x100001a] Thread (SaveData) [0x01f1fee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, “/dev_hdd0/game/BLES01069/USRDIR/UNREALENGINE3/BINARIES/PS3/SAVE/PIC1.PNG” [1]
E {PPU[0x100001a] Thread (SaveData) [0x01f1fee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, “/dev_hdd0/game/BLES01069/USRDIR/UNREALENGINE3/BINARIES/PS3/SAVE/SND0.AT3” [1]
```

**Comment by raveskirza** [comment]
With this PR Tekken Tag Tournament 2 save data dialog keeps popping up on every boot saying the "save data is Corrupt" and asks to overwrite the save data. Even after overwriting in the next boot again it identifies as corrupted

image

Attached Logs:
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/1544402/RPCS3.log)

Master does not identify the save data as corrupt

**Comment by jjsat** [comment]
Please try again with Tekken.

Skipping files not in PARAM.SFO was a mistake. I downloaded a save of Tekken Tag Tournament 2, decrypted it and replaced a save in the Wipeout HD demo with it. It did not enumerate any files which was clearly wrong.

Btw. Wipeout HD demo is pretty nice for checking this as it dumps information about the saves and callbacks to TTY without having to write your own game.

There is one noticable difference between this code and the real hardware, but it probably does not affect anything: sysSizeKB seems to be always 0x23 even if the save file does not exist yet and therefore there are no files in the directory. Also system files might not be included in the size count for savedata_get_list_item().

**Comment by Kravickas** [comment]
GTA IV doesnt work for me, idk what magic u used.

**Comment by MarioSonic2987** [comment]
@Kravickas Could you delete this folder or similar if it exists? Let the game create a new one. I did that.
```
dev_hdd0\home\00000001\savedata\BLES00229PROFILE
```

**Comment by SakataGintokiYT** [comment]
**savedata : 1e2e29b**
bez tytulu

**master:**
`F {PPU[0x1000008] Thread (LoadJM) [0x010c51b0]} MEM: Access violation reading location 0x0`

**Comment by Kravickas** [comment]
ok i am ingame GTA IV
and Tomb raider autosaves was fixed too

**Comment by raveskirza** [comment]
@jjsat Nice, tekken now no longer says save data is corrupt

**Comment by jjsat** [comment]
@SakataGintokiYT Check again with Thor. It should now be able to find existing game settings and allow to go ingame. The game seems to check for the existence of any save data with cellSaveDataUserListAutoLoad(), but does not want any actual data returned so it sets setBuf->dirListMax to 0. Handling of this case in the emulator is corrected in the last commit.

The game is still not able to proceed when trying to save. This is a different issue.

File: rpcs3/Emu/Cell/Modules/cellSaveData.cpp [hunk 1]
```diff
@@ -120,10 +120,11 @@ static NEVER_INLINE s32 savedata_op(ppu_thread& ppu, u32 operation, u32 version,
...
 				{
 					// Count the amount of matches and the amount of listed directories
-					if (listGet->dirListNum++ < setBuf->dirListMax)
+					listGet->dirNum++; // total number of directories
+					if (listGet->dirListNum < setBuf->dirListMax)
 					{
-						listGet->dirNum++;
-
+						listGet->dirListNum++; // number of directories in list
+
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellSaveData.cpp]: what the heck is this :)

Comment (jjsat) [reply rpcs3/Emu/Cell/Modules/cellSaveData.cpp]: A bunch of whitespace. Sorry I did not see it locally in my diff.

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellSaveData.cpp]: no worries. I just think it's funny

**Comment by AniLeo** [comment]
Can you rebase with fixed whitespaces?

**Comment by jjsat** [comment]
Rebased and whitespace should be fixed. I thought I could find a complete fix for the Thor save issue (and Wolfenstein has a similar one - it hangs while loading) but this might actually be a race condition. The emulator is a lot faster with the save functions and the scheduling is different.

**Comment by jjsat** [comment]
It is a race condition for both Thor (saving from menu) and Wolfenstein (loading).

Add this to the bottom of "savedata_op()" to wait a second before returning and all is good:
```
  thread_ctrl::spawn("Waiting Thread", [&]()
  {
    _sleep(1000);
    lv2_obj::awake(ppu);
  });

  lv2_obj::sleep(ppu);

  while (!ppu.state.test_and_reset(cpu_flag::signal))
  {
    thread_ctrl::wait();
  }
```

I am not including this in the PR because I don't know if this is the right way to do it. The technique is copied from sys_net.

**Comment by jarveson** [comment]
Out of curiosity, does kicking the callbacks to trigger through sysutil get past the race condition? Something like 

```cpp
atomic_t<bool> wait(true);
sysutil_register_cb([&](ppu_thread& ppu) -> s32
{
	func(ppu, CELL_MUSIC2_EVENT_INITIALIZE_RESULT, vm::make_var<s32>(CELL_OK), userData);
	wait = false;
	return CELL_OK;
});

while (wait) {
	thread_ctrl::wait_for(1);
}

```
around each one

**Comment by jjsat** [comment]
@jarveson That code never calls funcFile and just causes a freeze.

The reason why I did the more complicated lv2_obj::sleep waiting is because thread_ctrl::wait_for does not work in this situation. I have not looked into the scheduling behaviour of the emu in detail but in both games thread_ctrl::wait_for adds a delay but after that goes on as if no wait had happened. So it does obviously not reschedule as needed here.

Here is the behaviour for Wolfenstein:
- with thread_ctrl::wait_for: save is selected - savedata_op() is called - wait is called -> screen freezes for a second - savedata_op() returns - games shows "Reading save data" message on screen and hangs forever
- with lv2_obj::sleep: save is selected - savedata_op() is called - wait is called - games shows "Reading save data" and does not freeze - savedata_op() returns - game starts loading the map

Here it is for Thor:
- with thread_ctrl:wait_for: savedata_op() is called - wait is called -> game continues but does not return to main menu
- with lv2_obj::sleep: savedata_op() is called - wait is called -> game continues - savedata_op() returns - game goes to main menu

Also whie loading the save data during startup Thor displays a black screen on a PS3. With the lv2_obj::sleep patch it does this too. With thread_ctrl::wait_for it does not switch to black at all.

It surely would be better to delay piecewise where the PS3 would, so on every file access and not once at the end.

**Comment by jarveson** [comment]
Sorry, i guess i didnt make it clear, replace like `funcFile(ppu, result, fileGet, fileSet);` in the code i posted with the other callbacks, i dont know exactly what the games called so i just gave an example so ill spell it out.... 
so starting from the top, try replacing 
`funcList(ppu, result, listGet, listSet);`
with 
```cpp
atomic_t<bool> wait(true);
sysutil_register_cb([&](ppu_thread& ppu) -> s32
{
	funcList(ppu, result, listGet, listSet)
	wait = false;
	return CELL_OK;
});

while (wait) {
	thread_ctrl::wait_for(1);
}
```

`funcDone(ppu, result, doneGet);
`
with
```cpp
atomic_t<bool> wait(true);
sysutil_register_cb([&](ppu_thread& ppu) -> s32
{
	funcDone(ppu, result, fileGet, fileSet);
	wait = false;
	return CELL_OK;
});

while (wait) {
	thread_ctrl::wait_for(1);
}
```

`funcFixed(ppu, result, listGet, fixedSet);`
with

```cpp
atomic_t<bool> wait(true);
sysutil_register_cb([&](ppu_thread& ppu) -> s32
{
	funcFixed(ppu, result, listGet, fixedSet);
	wait = false;
	return CELL_OK;
});

while (wait) {
	thread_ctrl::wait_for(1);
}
```

`funcStat(ppu, result, statGet, statSet);` 
with
```cpp
atomic_t<bool> wait(true);
sysutil_register_cb([&](ppu_thread& ppu) -> s32
{
	funcStat(ppu, result, statGet, statSet);
	wait = false;
	return CELL_OK;
});

while (wait) {
	thread_ctrl::wait_for(1);
}
```

and `funcFile(ppu, result, fileGet, fileSet);`
with 
```cpp
atomic_t<bool> wait(true);
sysutil_register_cb([&](ppu_thread& ppu) -> s32
{
	funcFile(ppu, result, fileGet, fileSet);
	wait = false;
	return CELL_OK;
});

while (wait) {
	thread_ctrl::wait_for(1);
}
```

and see if it gets past the race condition...

**Comment by jjsat** [comment]
Yes, I understand that. I replaced the call to funcFile with your code and it does not call the callback function at all (disable optimization, put a breakpoint on the function call, no hit). The thread itself waits in the "while (wait)" loop forever.

**Comment by Kravickas** [comment]
can u rebase ?

**Comment by jjsat** [comment]
@SakataGintokiYT Added a fix for ArcaniA. The games now reaches the main menu. If no savegame is found on startup the game expects the call to cellSaveListAutoLoad to fail. It has to fail because the game returns 2 as the result of the funcFixed callback. I have not checked the other save functions that use the funcFixed callback, but there is actually a comment in the code that indicates that CELL_SAVEDATA_CBRESULT_OK_LAST_NOCONFIRM is not a valid return value for all callbacks.

Here is some [test code](https://gist.github.com/jjsat/10856b7d81990bdf3f99c8ddf9bca8b9).

**Comment by jjsat** [comment]
Is this new since the last commit? Do you have a log?

**Comment by Megamouse** [comment]
can you rebase first?

**Comment by Megamouse** [comment]
It starts with b9dfb2b

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1713287/RPCS3.log.gz)

**Comment by Nekotekina** [comment]
I would restrain from touching callback results at this moment.

**Comment by tamaskenta** [comment]
moon diver save problem.
E {PPU[0x1000005] Thread (FIO) [0x00720ee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, �/dev_hdd0/game/NPUB30345/USRDIR/BINARY/PS3/SND/INT/STREAM/BGM45_GAMEOVER.XWV� [1]
E {PPU[0x1000005] Thread (FIO) [0x00720ee4]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, �/dev_hdd0/game/NPUB30345/USRDIR/BINARY/PS3/SND/INT/STREAM/BGM01_TITLE.XWV� [2]
E {PPU[0x1000000] Thread (main_thread) [0x007242c0]} 'sys_fs_fcntl' failed with 0x80010006 : CELL_ENOENT, �/dev_hdd0/NPUB30345� [3]
E {PPU[0x1000000] Thread (main_thread) [0x00724308]} 'sys_fs_stat' failed with 0x80010006 : CELL_ENOENT, �/dev_hdd0/NPUB30345� [3]
E {rsx::thread} RSX: Swapchain: Swapchain creation failed because dimensions cannot fit. Max = 0, 0, Requested = 720, 576

**Comment by tamaskenta** [comment]
Game ID
NPUB30345.
PLEASE BUG FIXED CODE,THANK

**Comment by tamaskenta** [comment]
Not create data,bug((

**Comment by tamaskenta** [comment]
Not create saved data

**Comment by Nicba1010** [comment]
@tamaskenta please do not spam things here...

**Comment by AniLeo** [comment]
GitHub not allowing me to delete the spam comments for some reason..

**Comment by Asinin3** [comment]
Also fixes saving in Hellboy, probably need to check Rune Factory games at some point.

**Comment by uaqlover** [comment]
any update

**Pull request merged**